app/cdap/components/RecipeManagement/CreateRecipe/index.tsx

```diff
@@ -0,0 +1,182 @@
+      formData.recipeName?.trim().length === 0 ||
+      formData.description?.trim().length === 0 ||
+      nameErrorData.isRecipeNameError
+    ) {
+      setIsSaveDisabled(true);
+    } else {
+      setIsSaveDisabled(false);
+    }
+  };
+
+  const handleRecipeFormData = (formData: IRecipeFormData) => {
+    setRecipeFormData(formData);
+    handleSaveButtonMode(formData);
+  };
+
+  const onRecipeNameChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    handleRecipeFormData({
+      ...recipeFormData,
+      recipeName: event.target.value,
+    });
+    validateRecipeNameExists.current({
+      recipeName: event.target.value,
+      description: recipeFormData.description,
+    });
+  };
+
+  const onRecipeDescriptionChange = (event: React.ChangeEvent<HTMLTextAreaElement>) => {
+    handleRecipeFormData({
+      ...recipeFormData,
+      description: event.target.value,
+    });
+  };
+
+  const onFormSubmit = (event: FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+    onRecipeDataSave(recipeFormData);
+  };
+
+  const onRecipeDataSave = (recipeFormData: IRecipeFormData) => {
+    const requestBody = {
+      recipeName: recipeFormData.recipeName,
+      description: recipeFormData.description,
+      directives: recipeSteps,
+    };
+    createRecipeService({ requestBody, onCreateRecipeResponse, onCreateRecipeError });
+  };
+
+  const onCreateRecipeResponse = () => {
+    setShowRecipeForm(false);
+    setSnackbar({
+      open: true,
+      isSuccess: true,
+      message: `${recipeSteps.length} ${T.translate(`${PREFIX}.recipeSaveSuccessMessage`)}`,
+    });
+  };
+
+  const onCreateRecipeError = (err: Record<string, unknown>) => {
+    setShowRecipeForm(false);
+    setSnackbar({
+      open: true,
+      isSuccess: false,
+      message: (err.response as Record<string, string>).message,
+    });
+  };
+
+  const onCancel = () => {
+    setShowRecipeForm(false);
+  };
+
+  const onGetRecipeByNameError = (err: Record<string, unknown>, formData: IRecipeFormData) => {
+    if (err.statusCode === 404) {
+      setRecipeNameErrorData(noErrorState, formData);
+    }
+  };
+
+  const onGetRecipeByNameResponse = (formData: IRecipeFormData) => {
+    !recipeNameErrorData.isRecipeNameError &&
+      setRecipeNameErrorData(
+        {
+          isRecipeNameError: true,
+          recipeNameErrorMessage: T.translate(`${PREFIX}.sameNameErrorMessage`).toString(),
+        },
+        formData
+      );
+  };
+
+  const validateRecipeNameExists = useRef(
+    debounce((formData: IRecipeFormData) => {
+      if (formData.recipeName && !recipeNameRegEx.test(formData.recipeName)) {
+        setRecipeNameErrorData({
+          isRecipeNameError: true,
+          recipeNameErrorMessage: T.translate(`${PREFIX}.validationErrorMessage`).toString(),
+        });
+      } else {
+        if (formData.recipeName) {
+          getRecipeByNameService({ formData, onGetRecipeByNameResponse, onGetRecipeByNameError });
```

Comment:
Why create these `*Service` functions? Why not just subscribe to the Observable here?

Reply:
@njbriggs  @seanfeld  as per comment given by @radhikav1 we are maintaining common seperate file for services where we've defined all the APIs that are being used in Recipe Management, please suggest if changes needed

Reply:
This is fine for now. In general, though, using the Observable gives you a lot more options, especially when executing multiple operations in sequence or parallel.

Reply:
There is a weird state/callback flow here - this stuff either needs to be done in redux, using selectors and triggering actions that trigger api requests or in custom hooks. All these callbacks are very messy and hard to follow.

Reply:
This is not resolved

Reply:
used custom hook for API requests